game-core/build.install4j

```diff
@@ -34,7 +34,7 @@
       <dirEntry mountPoint="27" file="dice_servers" overwriteMode="4" shared="false" fileMode="644" uninstallMode="0" overrideFileMode="false" overrideOverwriteMode="false" overrideUninstallMode="false" entryMode="direct" subDirectory="dice_servers" excludeSuffixes="" dirMode="755" overrideDirMode="false">
         <exclude />
       </dirEntry>
-      <fileEntry mountPoint="22" file="./build/rootFiles/game_engine.properties" overwriteMode="4" shared="false" fileMode="644" uninstallMode="0" overrideFileMode="false" overrideOverwriteMode="false" overrideUninstallMode="false" />
+      <fileEntry mountPoint="22" file="game_engine.properties" overwriteMode="4" shared="false" fileMode="644" uninstallMode="0" overrideFileMode="false" overrideOverwriteMode="false" overrideUninstallMode="false" />
```

Comment:
Please note we do the version overwrite in travis with:
```- ENGINE_VERSION="$(grep engine_version game-core/game_engine.properties | sed 's/.*= *//g').$TRAVIS_BUILD_NUMBER"```

That modification is done before we build release artifacts, so the version number should be updated appropriately. I saw this in testing where I had to set the engine_number.

Reply:
Oh, I am mistaken, there is no `-i` in that sed command.

I see this is the proper block for the version update:
```
task prepareGameEngineProperties() {
    group = 'release'
    description = 'Updates the game engine properties with final values for distribution.'

    doLast {
        copy {
            from gameEnginePropertiesFile
            into gameEnginePropertiesArtifactFile.parent
        }
        ant.propertyfile(file: gameEnginePropertiesArtifactFile) {
            entry key: 'engine_version', value: version
        }
    }
}
```

This might be actually getting to where the build bug is; the installer prepare task does not directly depend on 'gameEngineProperties', we get that after buliding the all platforms. The build script seems to have assumed all platforms will always be built first before doing the installers, even though there is no requirement or dependency for that ordering. Hence if we run parallel builds, we can see the hypothesized race condition.

------------

I really like the idea of simply using the file from source location; but we would wind up modifying it and that is not ideal for a local build. We could solve that by injecting the full version number via variable, but having it checked in, at least the version number prefix seems useful for now.

I'll put this into WIP while I look at this.
- seems like the original location was okay in the end
- missing dependency on the engine props file task seems to have been the real problem.

Reply:
I've updated the travis build script to update the engine number in-place. Gradle then no longer needs to do the property file copy, so I was able to remove that. With that task removed, the gradle task dependency chain got a bit more linear and was simplified a bit further.

Reply:
It seems that having a local Gradle build and the Travis Gradle build produce identical artifacts would be desirable.  The goal of #2085 was to ensure that all artifact build logic was encapsulated in the Gradle build script rather than being split between Gradle and Travis scripts.

Reply:
Notice that in #2085 we still do a grep to get the 'base' of the engine version number. That is a key difference for as long as we did that, we could not put a '.dev' suffix in the game property file directly.

> being split between Gradle and Travis scripts.

Responding to this directly, perhaps interesting, gradle now has zero responsibility to the version number. Travis already owned the 'build number', it would always in some way have to know how that is injected to the version. 

Also of note, travis already also owned the lobby build number; having it own all of the final build numbers seems reasonable at the moment. We've also broken the dependency of the lobby version on game engine, as previously we got the 'root' version number from game engine and then used that as lobby. 

That's at least the rationale for the update at the moment, FWIW.

Reply:
> If you don't update game-core/game_engine.properties in parallel with latest_version.properties

this was and is a problem before and after. 


> Having a single source of truth for the engine version would seem to be preferable. In that respect, this change moves further away from that goal because now we have to update three files instead of two.

I respectfully disagree. Before we had to keep the properties file, gradle and travis all in sync. Now travis owns the full version number and writes it on build. We could release successfully without updating the game_engine.properties file, and `latest_version.properties` is the same as it always was.

> Ideally, one shouldn't even have to manually update latest_version.properties during a version bump. 

yes, but we don't want to notify users to upgrade their engine any time we do any new build. This mechanism is how we 'force UI notifications'.

The thing to really be sure is that before we had this unholy mixture where travis read the engine number from game code, it then appended and overwrote it in a compiled file. In a way travis owned a good chunk of the build number here, so by moving more responsibilty we did not increase burden. Meanwhile, gradle is now blissfully unaware of version details, it always uses exactly what is in properties. This will make our local testing more accurate as there are fewer moving parts and no need to simulate the build number append in an artificial way.

Reply:
Couple more points to consider:
- if we move the game_engine.properties file, travis would break, this shows the non-obvious coupling between the two (which is similar to `latest_version.properties`, we had to keep all in sync).
- lobby version should not be owned by game-core from `game_engine.properties`. By reading the 'base' version from the game_engine, we coupling lobby and game version, a bad thing. 
- fundamentally travis owns the build number as it is the thing that gives us build number. The fewer places we wire that to, the better. Making gradle not need or be aware of how to manipulate version number is a small win.

Reply:
With apologies for piling on, I'm looking at the complexity of the version number relative to where it is defined and how it evolves as we do a build. Notably, now the version number flows linearly in the system, before it was not a linear flow, and was a bit complex to understand. The extra touch point in travis is a very minor problem, in some ways a feature to me as we cannot forget to update that file, and as long as we get that one update, we will successfully update the build number.